pkg/internal/controller/controller_test.go

```diff
@@ -122,6 +122,26 @@ var _ = Describe("controller", func() {
 			close(done)
 		})
 
+		It("should error when cache sync timeout occurs", func(done Done) {
+			ctrl.CacheSyncTimeout = 10 * time.Nanosecond
+			c, err := cache.New(cfg, cache.Options{})
+			Expect(err).NotTo(HaveOccurred())
+			_, err = c.GetInformer(context.TODO(), &appsv1.Deployment{})
+			Expect(err).NotTo(HaveOccurred())
+			sync := false
+			ctrl.startWatches = []watchDescription{{
+				src: source.NewKindWithCache(&appsv1.Deployment{}, &informertest.FakeInformers{
```

Comment:
Instead of using the `FakeInformer`, please use the cache you construct in line 127 (The testcase you copied this from seems to be bugged, it constructs a cache and then never uses it).

To test the positive case (we correctly sync when there is a reasonable timeout), you could do the following:
* Start off similiar like this test, construct a cache and set it up in `startWatches` 
* Build a wrapper struct for the `Source` that closes a channel after `WaitForSync` was called (below)
* Start the cache in a distinct goroutine (as that is blocking)
* Start the controller in a distinct goroutine (as that is blocking, too)
* Wait for the channel in the source or a timeout via `select`
* Close the context used to start the cache and controller

wrapper would be something like this:
```
type signalingSourceWrapper struct {
  cacheSyncDone chan struct{}
  source.Source
}

func (s *singalingSourceWrapper) WaitForSync(ctx context.Context) error {
  defer func(){close(s.cacheSyncDone} ()
  return s.Source.WaitForSync(ctx)
}
```

Let me know if you have further questions on this

Reply:
ops, sorry, I had missed this. Does the current implementation look fine or shall I modify it ?

Reply:
I tried to do the following:
1. Created a wrapper around `SyncingSource` as that in turn wraps Source:
```
type SingnallingSourceWrapper struct {
	cacheSyncDone chan struct{}
	source.SyncingSource
}

func NewSignallingSource(cacheChannel chan struct{}, src source.SyncingSource) source.SyncingSource {
	return &SingnallingSourceWrapper{cacheSyncDone: cacheChannel, SyncingSource: src}
}

func (s *SingnallingSourceWrapper) WaitForSync(ctx context.Context) error {
	defer func() {
		close(s.cacheSyncDone)
	}()
	return s.SyncingSource.WaitForSync(ctx)
}
```

In the test, I tried the following as you have suggested:

```
// Configure a new cache:
c, err := cache.New(cfg, cache.Options{})
Expect(err).NotTo(HaveOccurred())
ctx := context.TODO()


// Set it up with start watches. Create a new `channel` for SignallingSource:
src := source.NewKindWithCache(&appsv1.Deployment{}, c)
ctrl.startWatches = []watchDescription{{
	src: NewSignallingSource(channel, src),
}}

// Start cache and controller in a separate go routine:
go c.Start(ctx)
go ctrl.Start(ctx)

// Wait for channel to close:
select {
	case <-channel:
		// channel has closed, cache has synced.
	default:
		// there is a timeout. cache did not sync
}

ctx.Done()
```

When I check for the condition where cache is synced, there is always a timeout. And I am also facing the error:
```
watch of *v1.Deployment ended with: very short watch: pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Unexpected watch close - watch lasted less than a second and no items received
```

I'm not able to figure out the reason for timing out always since the context which is being sent for starting caches and controller is closed in the end. Also the `channel` seems to still be open. Is there something which Im missing?

Reply:
Can you check the error return of the cache and controller?

Reply:
Also, the `default` case needs to not be `default` but something that actually times out, i.E.:

```
timeoutCtx, cancel := context.WithTimeout(context.Background(), 10 * time.Second)
defer cancel()
select {
	case <-channel:
		// channel has closed, cache has synced.
	case <-timeoutCtx.Done():
		// there is a timeout. cache did not sync
}
```

Reply:
The controller does return an error:
```
ErrStatus: {
                TypeMeta: {Kind: "", APIVersion: ""},
                ListMeta: {
                    SelfLink: "",
                    ResourceVersion: "",
                    Continue: "",
                    RemainingItemCount: nil,
                },
                Status: "Failure",
                Message: "Timeout: failed waiting for *v1.Deployment Informer to sync here",
                Reason: "Timeout",
                Details: {Name: "", Group: "", Kind: "", UID: "", Causes: nil, RetryAfterSeconds: 0},
                Code: 504,
            }
```

It seems to error while waiting for informer to sync (ie) [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/internal/informers_map.go#L187-L189)

Reply:
Ah. Well, that simplifies the whole thing, we can then simply start the controller synchronously and verify we get that error

Reply:
Yeah, the same error does occur when controller is started synchronously without any timeout set. But isn't informer not getting synced also mean that caches are not synced? if so, how can we verify the case that everything works fine and caches are synced when no timeout is set.

Reply:
So the reason this happens is that the cache isn't started properly by the time we start the controller. Since `cache.Start()` is blocking, it can not be used well to figure out when it finished starting. You could probably use `Eventually` with `cache.List()` for something that is not a deployment to verify that it finished starting before starting the controller (its important that its a different object than whatever the controller has, as the `List` creates and syncs an informer for that object type)

Reply:
When I start the cache in a separate go routine, wait for it to start, then start a controller in a separate go routine - I get the error `pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: watch of *v1.ReplicaSet ended with: very short watch: pkg/mod/k8s.io/client-go@v0.19.2/tools/cache/reflector.go:156: Unexpected watch close - watch lasted less than a second and no items received`

Why is it required that when we set a watch, there should be a change in the state of the resource ?

Reply:
Not sure why you are seeing that error, it is not required to have any change in the state of the resources for this to work. What works for me is this:
```
    It("should error when cache sync timeout occurs", func(done Done) {
      c, err := cache.New(cfg, cache.Options{})
      Expect(err).NotTo(HaveOccurred())
      ctrl.CacheSyncTimeout = 10 * time.Millisecond
      ctrl.startWatches = []watchDescription{{src: source.NewKindWithCache(&appsv1.Deployment{}, c)}}

      err = ctrl.Start(context.TODO())
      Expect(err).To(HaveOccurred())
      Expect(err.Error()).To(ContainSubstring("cache did not sync"))

      close(done)
    })

    It("should not error when cache sync time out is of reasonable value", func(done Done) {                                                                                                                                                                                                                                   
      ctx, cancel := context.WithCancel(context.Background())
      defer cancel()

      ctrl.CacheSyncTimeout = 1 * time.Second
      c, err := cache.New(cfg, cache.Options{})
      Expect(err).NotTo(HaveOccurred())

      sourceSynced := make(chan struct{})
      ctrl.startWatches = []watchDescription{{
        src: &syncingSourceWrapper{
          SyncingSource: source.NewKindWithCache(&appsv1.Deployment{}, c), 
          synced:        sourceSynced,
        },
      }}  

      // Start the cache and controller asynchronously and wait for the source to sync
      go func() {
        defer GinkgoRecover()
        Expect(c.Start(ctx)).To(Succeed())
      }() 
      go func() {
        defer GinkgoRecover()
        Expect(ctrl.Start(ctx)).To(Succeed())
      }() 
      <-sourceSynced

      close(done)
    }, 10.0)
```

Using this wrapper:
```
type syncingSourceWrapper struct {
  source.SyncingSource
  synced chan struct{}
}

func (ssw *syncingSourceWrapper) WaitForSync(ctx context.Context) error {
  if err := ssw.SyncingSource.WaitForSync(ctx); err != nil {
    return err
  }
  close(ssw.synced)
  return nil
} 
```